include/wil/com_apartment_variable.h

```diff
@@ -0,0 +1,243 @@
+            }
+        };
+
+        template<typename test_hook = apt_var_platform>
+        struct apartment_variable_base
+        {
+            inline static winrt::slim_mutex s_lock;
+
+            using variables_map = std::unordered_map<apartment_variable_base<test_hook>*, std::any>;
+
+            // Apartment id -> variable storage.
+            // Variables are stored using the address of the global variable as the key.
+            inline static std::unordered_map<unsigned long long, variables_map> s_apartmentStorage;
+
+            apartment_variable_base() = default;
+            ~apartment_variable_base()
+            {
+                clear();
+            }
+
+            // non-copyable, non-assignable
+            apartment_variable_base(apartment_variable_base const&) = delete;
+            void operator=(apartment_variable_base const&) = delete;
+
+            // get current value or throw if no value has been set
+            std::any& get_existing()
+            {
+                if (auto any = get_if())
+                {
+                    return *any;
+                }
+                THROW_HR(E_NOT_SET);
+            }
+
+            static variables_map* get_current_apartment_variables()
+            {
+                auto storage = s_apartmentStorage.find(test_hook::GetApartmentId());
+                if (storage != s_apartmentStorage.end())
+                {
+                    return &storage->second;
+                }
+                return nullptr;
+            }
+
+            variables_map* ensure_current_apartment_variables()
+            {
+                auto variables = get_current_apartment_variables();
+                if (variables)
+                {
+                    return variables;
+                }
+
+                struct ApartmentObserver : public winrt::implements<ApartmentObserver, IApartmentShutdown>
+                {
+                    void STDMETHODCALLTYPE OnUninitialize(unsigned long long apartmentId) noexcept override
+                    {
+                        // This code runs at apartment rundown so be careful to avoid deadlocks by
+                        // extracting the variables under the lock then release them outside.
+                        auto variables = [apartmentId]()
+                        {
+                            auto lock = winrt::slim_lock_guard(s_lock);
+                            return s_apartmentStorage.extract(apartmentId);
+                        }();
+                    }
+                };
+                test_hook::RegisterForApartmentShutdown(winrt::make<ApartmentObserver>().get());
+                return &s_apartmentStorage.insert({ test_hook::GetApartmentId(), {} }).first->second;
+            }
+
+            // get current value or custom-construct one on demand
+            std::any& get_or_create(const std::function<std::any()>& creator)

```

Comment:
Can avoid the overhead of `std::function` by using a custom lightweight functor.

```cpp
struct any_maker_base
{
    std::any (*adapter)(void*);
    void *inner;

    std::any operator()()
    {
        return adapter(inner);
    }
};

template<typename T>
struct any_maker : any_maker_base
{
    any_maker()
    {
        adapter = [](auto) -> std::any { return T{}; };
    }

    any_maker(T (*maker)())
    {
        adapter = [](auto maker) -> std::any { return reinterpret_cast<T(*)()>(maker)(); };
        inner = reinterpret_cast<void*>(maker);
    }

    template<typename F>
    any_maker(F&& f)
    {
        adapter = [](auto maker) -> std::any { return reinterpret_cast<F*>(maker)[0](); };
        inner = std::addressof(f);
    }
};
```

The `any_maker` also supports an arbitrary functor, which is handy in the case of apartment variables that are not global variables.

Then we get
```cpp
std::any& get_or_create(const any_maker& creator)
{
   ... as before ...
}
```

The `apartment_variable` methods would become

```cpp
        T& get_or_create()
        {
            return base::get_or_create(details::any_maker<T>());
        }

        // get current value or custom-construct one on demand
        template<typename F>
        T& get_or_create(F&& f)
        {
            return std::any_cast<T&>(base::get_or_create(details::any_maker<T>(std::forward<F>(f))));
        }
```